src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,314 @@ using namespace opossum::expression_functional;  // NOLINT
-  // (if a node has two leaves as inputs, it will be collected twice)
-  auto leaf_parents = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, LQPInputSide>>{};
-  visit_lqp(lqp, [&](auto& node) {
-    // Early out for non-leaves
-    if (node->input_count() != 0) {
-      return LQPVisitation::VisitInputs;
-    }
+void recursive_gather_required_expressions(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t>& outputs_visited_by_node) {
+  const auto additional_columns = gather_required_expressions(node);
+  auto& required_expressions = required_expressions_by_node[node];
+  required_expressions.insert(additional_columns.begin(), additional_columns.end());
+
+  ++outputs_visited_by_node[node];
+  if (outputs_visited_by_node[node] < node->output_count()) return;
 
-    auto pruned_column_ids = std::vector<ColumnID>{};
-    for (const auto& expression : node->column_expressions()) {
-      if (referenced_columns.find(expression) != referenced_columns.end()) {
-        continue;
+  // Once all nodes that may require columns from this node (i.e., this node's outputs) have been visited, we can
+  // recurse into this node's inputs.
+  for (const auto& input : {node->left_input(), node->right_input()}) {
+    if (!input) continue;
+
+    // Make sure the entry exists, then insert all expressions that the current node needs
+    required_expressions_by_node[input];
+    for (const auto& required_expression : required_expressions) {
+      // Add the columns needed here (and above) if they come from the input node
+      if (input->find_column_id(*required_expression)) {
+        required_expressions_by_node[input].emplace(required_expression);
       }
-      const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
-      pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
     }
 
-    // We cannot create a Table without columns - since Chunks rely on their first column to determine their row count
-    if (pruned_column_ids.size() == node->column_expressions().size() && !pruned_column_ids.empty()) {
-      pruned_column_ids.pop_back();
+    recursive_gather_required_expressions(input, required_expressions_by_node, outputs_visited_by_node);
+  }
+}
+
+void prune_join_node(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node) {
+  // Sometimes, joins are not actually used to combine data but only to check the existence of a tuple in a second
+  // table. Example: SELECT c_name FROM customer, nation WHERE c_nationkey = n_nationkey AND n_name = 'GERMANY'
+  // These joins will be rewritten to semi joins. However, we can only do this if the join is on a unique/primary
```

Comment:
The comment is a little bit confusing. What can we only do for the unique joins? Throwing out columns or using semi joins?

Reply:
```
  // Sometimes, joins are not actually used to combine data but only to check the existence of a tuple in a second
  // table. Example: SELECT c_name FROM customer, nation WHERE c_nationkey = n_nationkey AND n_name = 'GERMANY'
  // If the join is on a unique/primary key column, we know that each value on the right side of the join will be
  // unique and can rewrite an inner join to a semi join.
```

Better?

Reply:
"combine data" > "combine relations"/tables ?
Maybe begin the second paragraph with "Furthermore, " to make clear, that this is an additional step, that might be added on top.